### PR TITLE
UI: Sidebar context menu addon API

### LINKED ISSUE #29101: Story Actions API
Revive https://github.com/storybookjs/storybook/pull/28739

Implement new Story Actions API:

```ts
experimental_updateActions(actionId, {
  'component--story-a': {
    id: 'run-component-tests',
    icon: PlayIcon,
    title: 'Run component tests',
    event: PROVIDER_ID + '_RUN'
  },
  'component--story-b': {
    id: 'run-component-tests',
    icon: PlayIcon,
    title: 'Run component tests',
    event: PROVIDER_ID + '_RUN'
  }
})
```

### LINKED ISSUE #29540: Use Story Actions API in `addon-test` to run focused tests

### PR DESCRIPTION
**Pull request by ndelangen**
- Works on #29529
- Closes #29107
- Closes #29101
- Closes #29540

## What I did

- [x] allow custom link in menu list component
- [x] add a new method on the testProvider addon type for adding a contextMenu rendering
- [x] adjust the `Tree` component to render a context menu, including statuses & contextMenu renderings from testProviders
- [x] hoist the state of the sidebarBottom into a state module in `manager-api`

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

I'll be adding this section as soon as we have alignment on this feature's implementation

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  78.4 MB | 78.4 MB | 0 B | 0.35 | 0% |
| initSize |  144 MB | 144 MB | 167 kB | **1.68** | 0.1% |
| diffSize |  65.2 MB | 65.3 MB | 167 kB | **99.22** | 0.3% |
| buildSize |  6.88 MB | 7.03 MB | 156 kB | **97.52** | 2.2% |
| buildSbAddonsSize |  1.51 MB | 1.51 MB | 0 B | - | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.9 MB | 2.06 MB | 156 kB | **1788.27** | 🔺7.6% |
| buildSbPreviewSize |  271 kB | 271 kB | 0 B | - | 0% |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  3.88 MB | 4.04 MB | 156 kB | **1788.27** | 3.9% |
| buildPreviewSize |  3 MB | 3 MB | 148 B | 0.99 | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  22.8s | 6.7s | -16s -124ms | **-1.31** | 🔰-240.2% |
| generateTime |  19.2s | 21.2s | 2s | -0.05 | 9.5% |
| initTime |  13.1s | 15.5s | 2.3s | 0.33 | 15.3% |
| buildTime |  9.2s | 8.2s | -941ms | -0.66 | -11.3% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  5.8s | 5.5s | -254ms | -0.03 | -4.6% |
| devManagerResponsive |  3.7s | 3.3s | -413ms | -0.36 | -12.3% |
| devManagerHeaderVisible |  549ms | 613ms | 64ms | 0.89 | 10.4% |
| devManagerIndexVisible |  805ms | 686ms | -119ms | 0.73 | -17.3% |
| devStoryVisibleUncached |  863ms | 1.1s | 319ms | 1.23 | 27% |
| devStoryVisible |  804ms | 683ms | -121ms | 0.84 | -17.7% |
| devAutodocsVisible |  467ms | 519ms | 52ms | 0.38 | 10% |
| devMDXVisible |  481ms | 502ms | 21ms | 0.21 | 4.2% |
| buildManagerHeaderVisible |  495ms | 540ms | 45ms | -0.2 | 8.3% |
| buildManagerIndexVisible |  508ms | 556ms | 48ms | -0.22 | 8.6% |
| buildStoryVisible |  498ms | 534ms | 36ms | -0.27 | 6.7% |
| buildAutodocsVisible |  522ms | 446ms | -76ms | -0.46 | -17% |
| buildMDXVisible |  417ms | 416ms | -1ms | -0.36 | -0.2% |

## Greptile Summary

Added context menu functionality to Storybook's sidebar, allowing addons to inject custom links and actions through a new experimental test provider API.

- Added `contextMenu` property to `Addon_TestProviderType` in `code/core/src/types/modules/addons.ts` for custom menu items
- Implemented `useContextMenu` hook in `code/core/src/manager/components/sidebar/Tree.tsx` to manage hover/click states
- Added test provider state management in `code/core/src/manager-api/modules/experimental_testmodule.ts`
- Modified `TooltipLinkList` in `code/core/src/components/components/tooltip/TooltipLinkList.tsx` to support custom content rendering
- Added example implementation in test addon (`code/addons/test/src/manager.tsx`) for running tests via context menu

### CODE REVIEW

File: code/core/src/components/components/tooltip/TooltipLinkList.tsx [hunk 1]
```diff
@@ -25,15 +26,27 @@ const Group = styled.div(({ theme }) => ({
...
-interface ItemProps extends Link {
+export type Link = CustomLink | NormalLink;
+
+/**
+ * This is a custom link that can be used in the `TooltipLinkList` component. It allows for custom
+ * content to be rendered in the list; it does not have to be a link.
+ */
+interface CustomLink {
+  id: string;
+  icon?: any;
```
Comment (ndelangen) [inline code/core/src/components/components/tooltip/TooltipLinkList.tsx:45]: revisit this

File: code/core/src/manager-api/modules/experimental_testmodule.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  updateTestproviderState(id: TestProviderId, update: Partial<TestProviderState>): void;
+  clearTestproviderState(id: TestProviderId): void;
+  runTestprovider(id: TestProviderId): void;
+  cancelTestprovider(id: TestProviderId): void;
+};
+
+export const init: ModuleFn = ({ store, fullAPI }) => {
+  let sessionState: TestProviders = {};
+  try {
+    sessionState = JSON.parse(sessionStorage.getItem(STORAGE_KEY) || '{}');
```
Comment (ndelangen) [inline code/core/src/manager-api/modules/experimental_testmodule.ts:40]: is this needed?

File: code/core/src/manager/components/sidebar/Tree.tsx [hunk 3]
```diff
@@ -638,3 +746,26 @@ export const Tree = React.memo<{
     </StatusContext.Provider>
   );
 });
+
+function generateTestProviderLinks(testProviders: TestProviders, context: API_HashEntry): Link[] {
+  return Object.entries(testProviders)
+    .map(([k, e]) => {
```
Comment (ndelangen) [inline code/core/src/manager/components/sidebar/Tree.tsx:752]: rename `k` & `e`

File: code/core/src/components/components/tooltip/WithTooltip.tsx [hunk 4]
```diff
@@ -133,7 +133,7 @@ const WithTooltipPure = ({
       {/* @ts-expect-error (non strict) */}
       {typeof tooltip === 'function' ? tooltip({ onHide: () => onVisibleChange(false) }) : tooltip}
     </Tooltip>
-  );
+  ) : null;
```
Comment (ndelangen) [inline code/core/src/components/components/tooltip/WithTooltip.tsx:136]: This is an optimization that makes a lot of sense to me...
Why call a function which we will not use the returned JSX of?

File: code/core/src/manager/components/sidebar/SidebarBottom.tsx [hunk 5]
```diff
@@ -126,45 +122,14 @@ export const SidebarBottomBase = ({
...
-        cancelling: false,
-        running: true,
-        failed: false,
-        crashed: false,
-        progress: undefined,
-      });
+      clearTestproviderState(providerId);
       api.experimental_updateStatus(providerId, (state = {}) =>
         Object.fromEntries(Object.keys(state).map((key) => [key, null]))
       );
```
Comment (ndelangen) [inline code/core/src/manager/components/sidebar/SidebarBottom.tsx:130]: I'm considering moving this code into the state module as well.

File: code/core/src/manager/components/sidebar/SidebarBottom.tsx [hunk 6]
```diff
@@ -214,18 +179,18 @@ export const SidebarBottomBase = ({
...
+    api.on(TESTING_MODULE_PROGRESS_REPORT, onProgressReport);
 
     return () => {
-      api.getChannel()?.off(TESTING_MODULE_CRASH_REPORT, onCrashReport);
-      api.getChannel()?.off(TESTING_MODULE_PROGRESS_REPORT, onProgressReport);
-      api.getChannel()?.off(TESTING_MODULE_RUN_ALL_REQUEST, clearState);
+      api.off(TESTING_MODULE_CRASH_REPORT, onCrashReport);
+      api.off(TESTING_MODULE_PROGRESS_REPORT, onProgressReport);
+      api.off(TESTING_MODULE_RUN_ALL_REQUEST, clearState);
     };
```
Comment (ndelangen) [inline code/core/src/manager/components/sidebar/SidebarBottom.tsx:190]: I'm considering moving this code into the state module as well.

Right now, every time `testProviders` changes (which happens a LOT due to that's where the state is), we unsubscribe and resubscribe.

It's technically possible to miss events during this short period.

File: code/core/src/types/modules/addons.ts [hunk 7]
```diff
@@ -474,6 +477,13 @@ export interface Addon_TestProviderType<
   name: string;
   title: (state: Addon_TestProviderState<Details>) => ReactNode;
   description: (state: Addon_TestProviderState<Details>) => ReactNode;
+  contextMenu?: (
+    options: {
+      context: API_HashEntry;
+      state: Addon_TestProviderState<Details>;
+    },
+    components: { ListItem: typeof ListItem }
+  ) => ReactNode;
```
Comment (ndelangen) [inline code/core/src/types/modules/addons.ts:486]: This is the added API to `Addon_TestProviderType`.

I'm personally a bit on the fence about the `components` bit, as the user is likely to need more components anyway. Adding more is an option, but @JReinhold has stronger opinions on this than I do.

File: code/core/src/manager-api/modules/experimental_testmodule.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+        // TODO: get actual list and emit, this notification is for development purposes
+        fullAPI.addNotification({
+          id: 'testing-module',
+
+          content: {
+            headline: 'Running tests',
+            subHeadline: `Running tests for ${listOfFiles} stories`,
+          },
+        });
```
Comment (ndelangen) [inline code/core/src/manager-api/modules/experimental_testmodule.ts:85]: We need to implement the subset-testProvider running here @JReinhold

File: code/addons/test/src/RelativeTime.tsx
Comment (JReinhold) [inline code/addons/test/src/RelativeTime.tsx:1]: move to `src/components`

File: code/addons/test/src/components/ContextMenuItem.tsx [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        timestamp={new Date(state.progress.finishedAt)}
+        testCount={state.progress.numTotalTests}
+      />
+    );
+  } else if (state.watching) {
+    description = 'Watching for file changes';
+  }
+
+  return (
+    <div onClick={(event) => event.stopPropagation()}>
```
Comment (JReinhold) [inline code/addons/test/src/components/ContextMenuItem.tsx:79]: why stopPropagation? let's add a comment

File: code/addons/test/src/components/Title.tsx [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  return (
+    <div>
+      <Spaced col={1}>
+        <span style={{ display: 'inline-block', verticalAlign: 'middle' }}>Component tests</span>
+        {interactionsCount && !hasException ? (
+          <Badge status="neutral">{interactionsCount}</Badge>
+        ) : null}
+        {hasException ? <Badge status="negative">{interactionsCount}</Badge> : null}
+      </Spaced>
+    </div>
```
Comment (JReinhold) [inline code/addons/test/src/components/Title.tsx:21]: div looks unnecessary

File: code/addons/test/src/manager.tsx [hunk 11]
```diff
@@ -58,26 +47,27 @@ export function getRelativeTimeString(date: Date): string {
...
+const Info = styled.div({
+  display: 'flex',
+  flexDirection: 'column',
+  marginLeft: 6,
+});
 
-      const interval = setInterval(() => {
-        setRelativeTimeString(getRelativeTimeString(timestamp).replace(/^now$/, 'just now'));
-      }, 10000);
+const Title2 = styled.div<{ crashed?: boolean }>(({ crashed, theme }) => ({
```
Comment (JReinhold) [inline code/addons/test/src/manager.tsx:56]: find a better name for Title2

File: code/addons/test/src/manager.tsx [hunk 12]
```diff
@@ -91,50 +81,102 @@ addons.register(ADDON_ID, (api) => {
...
       watchable: true,
-
       name: 'Component tests',
-      title: ({ crashed, failed }) =>
-        crashed || failed ? 'Component tests failed' : 'Component tests',
-      description: ({ failed, running, watching, progress, crashed, error }) => {
-        const [isModalOpen, setIsModalOpen] = useState(false);
 
-        const errorMessage = error?.message;
+      contextMenu: ({ context, state }, { ListItem }) => {
```
Comment (JReinhold) [inline code/addons/test/src/manager.tsx:86]: 1. handle stories without the `'test'` tag.
2. handle when clicking Roots/Dirs/Components, where no leaf stories have the `'test'`, and not show the menu

File: code/core/src/manager-api/modules/experimental_testmodule.ts [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+      };
+      return store.setState(
+        ({ testProviders }) => {
+          return { testProviders: { ...testProviders, [id]: { ...testProviders[id], ...update } } };
+        },
+        { persistence: 'session' }
+      );
+    },
+    runTestProvider(id, options) {
+      console.log('LOG: runTestProvider', id, options);
```
Comment (JReinhold) [inline code/core/src/manager-api/modules/experimental_testmodule.ts:79]: clean up logs

**Review by JReinhold — APPROVED** [review]

**Pull request merged**